controllers/gce/utils/utils_test.go

```diff
@@ -0,0 +1,482 @@
+		if rfn != fooTest.efn {
+			t.Errorf("returned %v but expected %v", rfn, fooTest.efn)
+		}
+	}
+}
+
+func TestNerNamer(t *testing.T) {
+	fooTests := []struct {
+		cn  string
+		fn  string
+		ecn string
+		efn string
+	}{
+		{"", "", "", ""},
+		{"fooCN", "fooFN", "fooCN", "fooFN"},
+		{"fooCN01--fooCN02--fooCN03", "fooCN01--fooCN02--fooCN03", "fooCN03", "fooCN01--fooCN02--fooCN03"},
+	}
+
+	for _, fooTest := range fooTests {
+		namer := NewNamer(fooTest.cn, fooTest.fn)
+		rcn := namer.GetClusterName()
+		if rcn != fooTest.ecn {
+			t.Errorf("returned %v but expected %v", rcn, fooTest.ecn)
+		}
+
+		rfn := namer.GetFirewallName()
+		if rfn != fooTest.efn {
+			t.Errorf("returned %v but expected %v", rfn, fooTest.efn)
+		}
+	}
+}
+
+func TestTruncate(t *testing.T) {
+	nameLenLimitValue := ""
+	for i := 0; i < nameLenLimit; i++ {
+		nameLenLimitValue = fmt.Sprintf("%s%s", nameLenLimitValue, "t")
+	}
+
+	fooTests := []struct {
+		k  string
+		ek string
+	}{
+		{"", ""},
+		{"fooKey", "fooKey"},
+		{nameLenLimitValue + "more", nameLenLimitValue + "0"},
+	}
+
+	for _, fooTest := range fooTests {
+		namer := NewNamer("fooCN", "fooFN")
+		rk := namer.Truncate(fooTest.k)
+		if rk != fooTest.ek {
+			t.Errorf("returned %v but expected %v", rk, fooTest.ek)
+		}
+	}
+}
+
+func TestDecorateName(t *testing.T) {
+	nameLenLimitValue := ""
+	for i := 0; i < nameLenLimit-6; i++ {
+		nameLenLimitValue = fmt.Sprintf("%s%s", nameLenLimitValue, "t")
+	}
+
+	fooTests := []struct {
+		cn  string
+		n   string
+		edn string
+	}{
+		{"", "", ""},
+		{"", "fooName", "fooName"},
+		{"fooCN", "fooName", "fooName--fooCN"},
+		{"fooCN", nameLenLimitValue, nameLenLimitValue + "--fooC0"},
+	}
+
+	for _, fooTest := range fooTests {
+		namer := NewNamer(fooTest.cn, "fooFN")
+		rdn := namer.decorateName(fooTest.n)
+		if rdn != fooTest.edn {
+			t.Errorf("returned %v but expected %v", rdn, fooTest.edn)
+		}
+	}
+}
+
+func TestParseName(t *testing.T) {
+	fooTests := []struct {
+		n   string
+		ecn string
+		ers string
+	}{
+		{"", "", ""},
+		{"foo", "", ""},
+		{"foo01--", "", ""},
+		{"foo01--foo02", "foo02", ""},
+		{"foo01-foo02--foo03", "foo03", "foo02"},
+	}
+
+	namer := NewNamer("fooCN", "fooFN")
+	for _, fooTest := range fooTests {
+		nc := namer.ParseName(fooTest.n)
+		if nc == nil {
+			t.Errorf("returned nil")
+			continue
+		}
+
+		if nc.ClusterName != fooTest.ecn {
+			t.Errorf("returned %v but expected %v", nc.ClusterName, fooTest.ecn)
+			continue
+		}
+
+		if nc.Resource != fooTest.ers {
+			t.Errorf("returned %v but expected %v", nc.Resource, fooTest.ers)
+			continue
+		}
+	}
+}
+
+func TestNameBelongsToCluster(t *testing.T) {
+	fooTests := []struct {
+		n  string
+		cn string
+		ev bool
+	}{
+		{"", "", false},
+		{"fooName", "", false},
+		{"k8s-fooName", "fooCN", false},
+		{"k8s-fooName", "", true},
+		{"k8s-foo01--foo02--foo03", "foo", false},
+		{"k8s-foo01--foo02", "foo02", true},
+		{"k8s-foo01--foo02", "", false},
+		{"k8s-foo01--foo02", "foo", false},
```

Comment:
i think a truncated name is not a special test case for this method, but i can not sure my understanding is right, so i added a test case about a truncated name. If my understanding is wrong, please give me some more descriptions,  thank you!